UnityProject/Storybook/Assets/Code/Scripts/Game/Player/PlayerMover.cs

```diff
@@ -0,0 +1,338 @@
+﻿using UnityEngine;
+using System.Collections;
+using System.Collections.Generic;
+
+/// <summary>
+/// Class for handling player movement.
+/// </summary>
+public class PlayerMover : NetworkMover
+{
+
+    private RoomObject m_currentRoom;
+    private RoomData m_currentRoomData;
+    private Location m_currentRoomLoc;
+    private List<NetworkNodeMover> m_worldPlayers = new List<NetworkNodeMover>();
+    private int m_leaderId = 1;
+    private bool m_areButtonsEnabled = false;
+
+    /// <summary>
+    /// Spawns all the players in the given room object
+    /// Called only to place the players in the starting room of the floor
+    /// </summary>
+    /// <param name="room">The room to spawn the players in</param>
+    public void SpawnInRoom(RoomObject room)
+    {
+        m_currentRoom = room;
+        m_currentRoomLoc = room.RoomLocation;
+        MapManager mapManager = FindObjectOfType<MapManager>();
+        m_currentRoomData = mapManager.GetRoomData(room.RoomLocation.X, room.RoomLocation.Y);
+
+        photonView.RPC("SendRoomData", PhotonTargets.Others, m_currentRoomData.IsNorthDoorActive, m_currentRoomData.IsEastDoorActive,
+            m_currentRoomData.IsSouthDoorActive, m_currentRoomData.IsWestDoorActive, m_currentRoomData.RoomType);
+        photonView.RPC("SendRoomLoc", PhotonTargets.Others, m_currentRoomLoc.X, m_currentRoomLoc.Y);
+
+        foreach(NetworkNodeMover player in m_worldPlayers)
+        {
+            player.IsAtTarget = true;
+        }
+
+        if (m_worldPlayers.Count >= 1)
+        {
+            m_worldPlayers[0].transform.position = room.Player1Node.transform.position;
+            m_worldPlayers[0].TargetNode = room.Player1Node;
+        }
+        if (m_worldPlayers.Count >= 2)
+        {
+            m_worldPlayers[1].transform.position = room.Player2Node.transform.position;
+            m_worldPlayers[1].TargetNode = room.Player2Node;
+        }
+        if (m_worldPlayers.Count >= 3)
+        {
+            m_worldPlayers[2].transform.position = room.Player3Node.transform.position;
+            m_worldPlayers[2].TargetNode = room.Player3Node;
+        }
+        if (m_worldPlayers.Count >= 4)
+        {
+            m_worldPlayers[3].transform.position = room.Player4Node.transform.position;
+            m_worldPlayers[3].TargetNode = room.Player4Node;
+        }
+
+        AreButtonsEnabled = true;
+    }
+
+    /// <summary>
+    /// Moves all of the players through the given door
+    /// It is a coroutine so that it can update the player positions once they reach the door
+    /// </summary>
+    /// <param name="door">The door for the players to go through</param>
+    /// <returns></returns>
+    public IEnumerator MoveThroughDoor(Door door)
```

Comment:
A cleaner design for this would be to have MoveThroughDoor return void (or a boolean). When it gets called it does one of two things. If the door is open then it starts a Courroutine to move through the door to the next room. If the door is closed then send off an event in the event dispatcher. You can create a new event dispatcher type for room events, and then create a method call like "OnOpenDoor" or something like that. The event can take in a delegate as a parameter which will allow a callback from other systems.